Lean4Lean/TypeChecker.lean

```diff
@@ -323,11 +353,13 @@ def whnfCore' (e : Expr) (cheapRec := false) (cheapProj := false) : RecM Expr :=
   | .bvar .. | .sort .. | .mvar .. | .forallE .. | .const .. | .lam .. | .lit ..
   | .mdata .. => unreachable!
   | .fvar _ => return ← whnfFVar e cheapRec cheapProj
-  | .app .. =>
+  | .app .. => -- beta-reduce at the head as much as possible, apply any remaining `rargs` to the resulting expression, and re-run `whnfCore`
     e.withAppRev fun f0 rargs => do
+    -- the head may still be a let variable/binding, projection, or mdata-wrapped expression
     let f ← whnfCore f0 cheapRec cheapProj
     if let .lam _ _ body _ := f then
       let rec loop m (f : Expr) : RecM Expr :=
+        -- FIXME only execute this when in one of the `else`s below (though the cost of `whnfCore` is minimal when `r` is a lambda)
```

Comment:
This should already be the case, although it's possible the `let r :=` functions are evaluated eagerly. `whnfCore` will definitely not be run eagerly, because this line creates a continuation and only runs it in the else branch. (Note that it is `let cont2 := do` and not `let cont2 <- do`, perhaps this is the source of the confusion.)